### PR TITLE
Convert the JSON code to use D slices instead of pointers

### PR DESCRIPTION
**Pull request by Geod24**
Since we want to get rid of strlen, I tried to convert a simple global to a D slice and was blocked by the JSON module. So here is the module conversion to D.
Some simple things were added (e.g. `RootObject.toString` which just `strlen` on `toChars`), but I avoided changing the AST as much as possible, so things like `kind()` and `Loc.filename` can be done in another PR.

### CODE REVIEW

File: src/dmd/json.d
Comment (WalterBright) [inline src/dmd/json.d]: Not too thrilled about this. It should be type.toString(). Yes, the C++ code cannot call it, but all you need to do is put a dummy function declaration there on the C++ side so the vtbl[] entries line up.

Comment (Geod24) [reply src/dmd/json.d]: Fixed

File: src/dmd/dmodule.d
Comment (JinShil) [inline src/dmd/dmodule.d]: DDoc?

File: src/dmd/hdrgen.d
Comment (JinShil) [inline src/dmd/hdrgen.d]: DDoc?

File: src/dmd/hdrgen.d
Comment (JinShil) [inline src/dmd/hdrgen.d]: DDoc?

File: src/dmd/hdrgen.d
Comment (JinShil) [inline src/dmd/hdrgen.d]: DDoc?

File: src/dmd/json.d
Comment (jacob-carlborg) [inline src/dmd/json.d]: I'm guessing you're trying keep the new to code be as close as possible to the old code but this first expression (`!this.filename.length`) isn't needed anymore?

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d]: As I mentioned before, toString() is used pervasively in dmd, and as long as it does what the other toString()'s do, adding a Ddoc for it is redundant.

**Comment by Geod24** [comment]
Updated according to feedback

**Review by JinShil — APPROVED** [review]

File: src/dmd/dmodule.d
Comment (ibuclaw) [inline src/dmd/dmodule.d]: Provide a human

File: src/dmd/hdrgen.d
Comment (ibuclaw) [inline src/dmd/hdrgen.d]: A human

File: src/dmd/hdrgen.d
Comment (ibuclaw) [inline src/dmd/hdrgen.d]: A human

File: src/dmd/hdrgen.d
Comment (ibuclaw) [inline src/dmd/hdrgen.d]: A human

File: src/dmd/json.d
Comment (ibuclaw) [inline src/dmd/json.d]: Wouldn't all non-visit members of this visitor be private? Ergo you can just add `extern(D):` at the top of this body rather than adding it for each individual function.

Think about it for future refactor.

File: src/dmd/root/outbuffer.d
Comment (ibuclaw) [inline src/dmd/root/outbuffer.d]: Seems like a leftover from a previous attempt here.

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/dmd/root/outbuffer.d
Comment (Geod24) [reply src/dmd/root/outbuffer.d]: It is, but also its own commit

File: src/dmd/json.d
Comment (Geod24) [reply src/dmd/json.d]: True that. At the moment I'm putting `extern(D)` right and left (got more slice-ification coming), but I also had weird bugs (compiler SEGV) being triggered in the past some I decided to go for verbose and simple. But when enough is converted, we surely can do that.

**Comment by Geod24** [comment]
Fixed typos

File: src/dmd/dmodule.d
Comment (rainers) [reply src/dmd/dmodule.d]: While I agree that a comment is not really necessary here, you have to add some ddoc comment to make the function appear in the documentation at all. Not sure if that is a good thing as you have to litter the code with empty comments this way. Is there an option to put all (public) symbols into the documentation?

Comment (Geod24) [reply src/dmd/dmodule.d]: For the record I added a (trivial) DDOC comment just for this reason (documentation)

File: src/dmd/json.d
Comment (wilzbach) [inline src/dmd/json.d]: Why not put this as a global function into `utils` or similar? It's used at other places too and doesn't really fit in `ToJsonVisitor`

File: src/dmd/hdrgen.d
Comment (wilzbach) [inline src/dmd/hdrgen.d]: Could use `toDString` from below.

File: src/dmd/dmodule.d
Comment (wilzbach) [inline src/dmd/dmodule.d]: Could use `toDString` from below.

File: src/dmd/json.d
Comment (wilzbach) [inline src/dmd/json.d]: Nit: another advantage of making `toDString` a global function is that you can do the more natural: `s.comment.toDString()`

File: src/dmd/root/rootobject.d
Comment (wilzbach) [inline src/dmd/root/rootobject.d]: Could use `toDString` from above.

File: src/dmd/json.d
Comment (Geod24) [reply src/dmd/json.d]: Didn't expect it to be so widely used :(
Also I'm wary that `dmd.utils` will be imported everywhere soon.
But let's give it a shot...

**Comment by Geod24** [comment]
Addressed @wilzbach 's comments

File: src/dmd/root/rootobject.d
Comment (Geod24) [reply src/dmd/root/rootobject.d]: Here I choose not to, as we want `rootobject` to depends on as little as possible.

**Review by JinShil — APPROVED** [review]
Still looking good.

**Comment by Geod24** [comment]
Since this has gotten a bit more reviews, and one might be wondering where I am going, I added a bit more diff than originally intended:
- Commit 0c91769 moves `DArray` from `globals.h` to a standalone header, because we don't want `object.h` importing `globals.h`
- I reworked `RootObject.toString` to be callable from C++ and added a test.

As for the end goal is, [it's to be able to do this](https://github.com/Geod24/dmd/commit/1a740a2e98294e6d5430659d4ca358d5f05bbdda#diff-3ea076fe941194ddccb18fc686689e4eL148) gradually.

File: src/tests/cxxfrontend.c
Comment (jacob-carlborg) [inline src/tests/cxxfrontend.c]: Does this work on Windows? I’ve been told D arrays are returned in registers while structs of this size are not on Windows.

File: src/dmd/utils.d
Comment (jacob-carlborg) [inline src/dmd/utils.d]: Why not `toString`?

Comment (Geod24) [reply src/dmd/utils.d]: `toString` is to provide a string representation, this is to convert a `const(char)*` to a `const(char)[]`.

File: src/tests/cxxfrontend.c
Comment (Geod24) [reply src/tests/cxxfrontend.c]: I should've known it was too good to be true.
I'll wait for the auto-tester to tell me what's the situation with Windows, but it fails on 32 bits.
So I guess I will take that test out for now :'(

Comment (Geod24) [reply src/tests/cxxfrontend.c]: Hum, the C++ frontend tests only run on POSIX anyway...

**Comment by Geod24** [comment]
@wilzbach : 
```
dmd/root/rootobject.d(66:29)[warn]: Methods 'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' are non-const.
```

What's the deal with that warning ?? `extern(C++)` classes don't inherit from `Object` so it doesn't make sense to require them to be `const`. And I currently cannot make it `const`.

File: src/tests/cxxfrontend.c
Comment (jacob-carlborg) [reply src/tests/cxxfrontend.c]: For reference: https://github.com/dlang/dmd/pull/8120#discussion_r200835776.

**Comment by wilzbach** [comment]
> What's the deal with that warning ?? extern(C++) classes don't inherit from Object so it doesn't make sense to require them to be const. 

Yeah that's the newly added Dscanner static code analysis check and the idea is that the checks slowly allow to modernize DMD / enforce some automatic code checking.
I guess the assumption behind this check was that "opEquals, opCmp, toHash, and toString are either const, immutable or inout" because they shouldn't be changing any data.

> And I currently cannot make it const.

tl;dr: in this case, add your module to this list:

https://github.com/dlang/dmd/blob/48726c875fbfe16fd93051570d149e893d2dc3dc/src/.dscanner.ini#L118

File: src/dmd/root/object.h
Comment (jacob-carlborg) [inline src/dmd/root/object.h]: Is it possible to implement this on the C++ and have the implementation assert?

**Comment by Geod24** [comment]
@wilzbach : Thanks, glad to see there's a way out of it!

File: src/dmd/root/object.h
Comment (Geod24) [reply src/dmd/root/object.h]: I don't think so.

**Comment by wilzbach** [comment]
@Geod24 I don't know exactly what's up with SemaphoreCI, but https://github.com/dlang/dmd/pull/8560 (once merged back to master) will hopefully help a bit.

**Comment by Geod24** [comment]
Rebased on master, let's see.

**Review by JinShil — APPROVED** [review]
Yup, I still like it.

**Review by wilzbach — APPROVED** [review]
Yeah same here. One small step forward (:

**Comment by Geod24** [comment]
Anything else ? Got a few changes depending on this PR (to be precise, on `toDString`).

**Comment by wilzbach** [comment]
I think this is just pending/blocked on Ian's change request. @ibuclaw are you okay with moving forward here?

**Pull request merged**
